docs/user-guide.md

```diff
@@ -0,0 +1,58 @@
+
+Usually, operators use `ServiceBinding` resource name (`.metadata.name`) as the bindings directory name, but the spec also provides a way to override that name through the `.spec.name` field. So, there is a chance for bindings name collision.  However, due to the nature of the volume mount in Kubernetes, the bindings directory will contain values from only one of the Secret resources.  This is a caveat of using the bindings directory name to look up the bindings.
+
+### Purpose of the type and the provider fields in the application projection
+
+Service Binding specification mandates `type` field and recommends `provider` field in the binding Secret resource.  Normally `type` field should be good enough to look up the bindings.  The provider field could be used where there are different providers for the same Provisioned Service type.  For example, if the type is `mysql`, the provider value could be `mariadb`, `oracle`, `bitnami`, `aws-rds`, etc.  When the application is reading the binding values, if necessary, the application could consider `type` and `provider` as a composite key to avoid ambiguity.  It will be helpful if an application needs to choose a particular provider based on the deployment environment.  In the deployment environment (`stage`, `prod`, etc.), at any given time, you need to ensure only one bindings projection exits for a given `type` or `type` & `provider` -- unless your application needs to connect to all the services.
+
+### Programming language specific library APIs
+
+The application can retrieve the bindings through the library available for your programming language of choice.  The language-specific APIs are encouraged to follow the pattern described here.  This is not an exhaustive list of APIs.  Consult your library API documentation to confirm the behavior.
+
+The language binding API can create separate functions to retrieve bindings. For example, in Go:
+
+```
+Bindings(_type string) []map[string]string
+BindingsWithProvider(_type, provider string) []map[string]string
+```
+
+Languages with function overloading can use a method with different parameters to retrieve bindings. For example, in Java:
+
+```
+public List<Binding> filterBindings(@Nullable String type)
+public List<Binding> filterBindings(@Nullable String type, @Nullable String provider)
+```
+
+(Example taken from [Spring Cloud Bindings](https://github.com/spring-cloud/spring-cloud-bindings))
+
+Since the APIs are returning more than one value, depending on your application need, you can choose to connect to the first entry or all of them, if required.
+
+### Environment Variables
+
+The spec also has support for projecting environment variables.  You can use the built-in language feature of your programming language of choice to read environment variables.  The container must restart to update the values of environment variables if there is a change in the Secret resource.
```

Comment:
```suggestion
The specification also has support for projecting binding values as environment variables.  You can use the built-in language feature of your programming language of choice to read environment variables.  The container must restart to update the values of environment variables if there is a change in the binding.
```

Reply:
Much like the list name collision above, I feel like this should be a failure-to-reconcile if env-var names collide.  /cc @arthurdm and @scothis